src/app_conn_pool.rs

```diff
@@ -0,0 +1,319 @@
+use crate::config::{AllowedFns, Configuration};
+use crate::{HcHttpGatewayError, HcHttpGatewayResult};
+use futures::future::BoxFuture;
+use holochain_client::{
+    AdminWebsocket, AppWebsocket, AuthorizeSigningCredentialsPayload, CellInfo, ClientAgentSigner,
+    ConductorApiError, ConnectRequest, GrantedFunctions, IssueAppAuthenticationTokenPayload,
+    Timestamp, WebsocketConfig,
+};
+use holochain_types::app::InstalledAppId;
+use holochain_types::websocket::AllowedOrigins;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::{Arc, RwLock};
+
+/// The origin that the gateway will use when connecting to Holochain app interfaces.
+pub const HTTP_GW_ORIGIN: &str = "hc-http-gw";
+
+/// A wrapper around an app websocket connection that includes state required to manage the
+/// connection.
+#[derive(Debug, Clone)]
+pub struct AppWebsocketWithState {
+    /// The app websocket connection.
+    pub app_ws: AppWebsocket,
+    /// The time at which the connection was opened.
+    pub opened_at: Timestamp,
+}
+
+/// A connection pool for app connections.
+///
+/// This is a pool in the sense that it manages multiple connections to Holochain app interfaces,
+/// but it will manage exactly one connection per installed app.
+#[derive(Debug, Clone)]
+pub struct AppConnPool {
+    configuration: Configuration,
+    cached_app_port: Arc<RwLock<Option<u16>>>,
```

Comment:
Is this like reserving an app port?

Reply:
It's all local, so the listing call should be fast. Fast enough that I'd prefer to not cache it, but I'm not opposed if you consider this important.

Reply:
It might be a little heavy on optimisation. I'm a little concerned about this running alongside Envoy which has a bad habit of allocating a silly number of app interfaces - but that is going away...

I'd like to keep this for now, it's trivial to remove.